include/libsemigroups/to-presentation.hpp

```diff
@@ -43,173 +47,207 @@ namespace libsemigroups {
-    std::transform(
-        p.alphabet().cbegin(), p.alphabet().cend(), new_alphabet.begin(), f);
-
-    WordOutput new_inverses;
-    new_inverses.resize(p.inverses().size());
-    std::transform(
-        p.inverses().cbegin(), p.inverses().cend(), new_inverses.begin(), f);
-
-    InversePresentation<WordOutput> result;
-    result.contains_empty_word(p.contains_empty_word());
-    result.alphabet(new_alphabet);
-    result.inverses(new_inverses);
+              && !std::is_same_v<WordOutput, WordInput>,
+          InversePresentation<WordOutput>>;
+#endif
 
-    WordOutput rel;
-    for (auto it = p.rules.cbegin(); it != p.rules.cend(); ++it) {
-      rel.resize(it->size());
-      std::transform(it->cbegin(), it->cend(), rel.begin(), f);
-      result.rules.push_back(rel);  // TODO std::move?
-      rel.clear();
-    }
-    return result;
-  }
-
-  // TODO required? It's not even correct so remove!
+  //! \ingroup to_presentation_group
+  //!
+  //! \brief Make an inverse presentation from a presentation
+  //!
+  //! Returns an inverse presentation with rules equivalent to those of the
+  //! input presentation, but over a normalised alphabet. If the alphabet of
+  //! \p p is \f$\{a_0, a_1, \dots, a_{n-1}\}\f$, then the alphabet of the
+  //! returned inverse presentation will be \f$\{0, 1, \dots, n-1, n, \dots,
+  //! 2n-1\}\f$, where the inverse of letter \f$i\f$ is the letter \f$i + n\,
+  //! (\text{mod }2n)\f$.
+  //!
+  //! \tparam Word the type of the words in the presentation.
+  //! \param p the input presentation.
+  //!
+  //! \returns A value of type `InversePresentation<WordOutput>`.
+  //!
+  //! \throws LibsemigroupsException if `p.validate()` throws.
   template <typename Word>
   InversePresentation<Word>
-  to_presentation(InversePresentation<Word> const& p) {
-    return p;
-  }
+  to_inverse_presentation(Presentation<Word> const& p);
 
+  //! No doc
+  // TODO(now) is this the correct way to prevent the nonsensical conversion of
+  // an inverse presentation to an inverse presentation of the same type?
+  // JDE wanted some way of preventing the user from calling
+  // to_inverse_presentation(Presentation<Word> const& p) in the case where p
+  // is, in fact, already an inverse presentation.
   template <typename Word>
   InversePresentation<Word>
-  to_inverse_presentation(Presentation<Word> const& p) {
-    InversePresentation<Word> result;
-    result.rules.insert(result.rules.end(), p.rules.cbegin(), p.rules.cend());
-    result.alphabet_from_rules();
-    result.contains_empty_word(p.contains_empty_word());
-    presentation::normalize_alphabet(result);
-    result.alphabet(2 * result.alphabet().size());
-    auto invs = result.alphabet();
-    std::rotate(invs.begin(), invs.begin() + invs.size() / 2, invs.end());
-    result.inverses(std::move(invs));
-    // result.validate();
-    return result;
+  to_inverse_presentation(InversePresentation<Word> const& ip) {
+    return ip;
   }
```

Comment:
As the comment says, is this the best way to prevent an `InversePresentation` from being converted to an `InversePresentation` as if it were a normal `Presentation`? I.e. prevent the addition of extra generators to create inverses from.

Reply:
The code snippet above suggests that the old functionality has been removed. It hasn't; that function still exists. The above is an overload that specialises for the `InversePresentation` subclass of `Presentation`.